tests/baselines/reference/derivedClassSuperProperties.js

```diff
@@ -0,0 +1,844 @@
+        return _this;
+    }
+    return Derived1;
+}(Base));
+var Derived2 = /** @class */ (function (_super) {
+    __extends(Derived2, _super);
+    function Derived2() {
+        var _this = this;
+        _this.prop = true;
+        _super.prototype.receivesAnything.call(_this, _this);
+        _this = _super.call(this) || this;
+        return _this;
+    }
+    return Derived2;
+}(Base));
+var Derived3 = /** @class */ (function (_super) {
+    __extends(Derived3, _super);
+    function Derived3() {
+        var _this = this;
+        _this.prop = true;
+        _super.prototype.receivesAnything.call(_this);
+        _this = _super.call(this, _this) || this;
+        return _this;
+    }
+    return Derived3;
+}(Base));
+var Derived4 = /** @class */ (function (_super) {
+    __extends(Derived4, _super);
+    function Derived4() {
+        var _this = this;
+        _this.prop = true;
+        _super.prototype.receivesAnything.call(_this, _this);
+        _this = _super.call(this, _this) || this;
+        return _this;
+    }
+    return Derived4;
+}(Base));
+var Derived5 = /** @class */ (function (_super) {
+    __extends(Derived5, _super);
+    function Derived5() {
+        var _this = _super.call(this) || this;
+        _this.prop = true;
+        _super.prototype.receivesAnything.call(_this);
+        return _this;
+    }
+    return Derived5;
+}(Base));
+var Derived6 = /** @class */ (function (_super) {
+    __extends(Derived6, _super);
+    function Derived6() {
+        var _this = _super.call(this, _this) || this;
+        _this.prop = true;
+        _super.prototype.receivesAnything.call(_this);
+        return _this;
+    }
+    return Derived6;
+}(Base));
+var Derived7 = /** @class */ (function (_super) {
+    __extends(Derived7, _super);
+    function Derived7() {
+        var _this = _super.call(this) || this;
+        _this.prop = true;
+        _super.prototype.receivesAnything.call(_this, _this);
+        return _this;
+    }
+    return Derived7;
+}(Base));
+var Derived8 = /** @class */ (function (_super) {
+    __extends(Derived8, _super);
+    function Derived8() {
+        var _this = _super.call(this, _this) || this;
+        _this.prop = true;
+        _super.prototype.receivesAnything.call(_this, _this);
+        return _this;
+    }
+    return Derived8;
+}(Base));
+var DerivedWithArrowFunction = /** @class */ (function (_super) {
+    __extends(DerivedWithArrowFunction, _super);
+    function DerivedWithArrowFunction() {
+        var _this = this;
+        _this.prop = true;
+        (function () { return _this; })();
+        _this = _super.call(this) || this;
+        return _this;
+    }
+    return DerivedWithArrowFunction;
+}(Base));
+var DerivedWithArrowFunctionParameter = /** @class */ (function (_super) {
+    __extends(DerivedWithArrowFunctionParameter, _super);
+    function DerivedWithArrowFunctionParameter() {
+        var _this = this;
+        _this.prop = true;

```

Comment:
Thanks @rbuckton, this explanation is really helpful. I _think_ the transformers are fixed in afc58e123b2ab53627ec1721b1849ff8507af860, by having the emitted constructors contain, in order:

1. Pre-`super()` code
2. The `super()` itself
3. Properties with initializers
4. Parameter properties
5. Post-`super()` code

The `_a = () => {` in those snippets have me stumped, though. Is there a reason the contents can't be placed where the `_a();` is?

The updated baselines seem generally "better" than before. Meaning: wWithin the existing test cases that result in semantic errors for `this` access before `super`, several previously resulted in a global `_this` being created; now, they correctly defer referring to `_this` until it's assigned to `_super.call(this) || this`. Is that... good?

Reply:
Ping @rbuckton - do you think this could make it in for 3.8?

Reply:
> The `_a = () => {` in those snippets have me stumped, though. Is there a reason the contents can't be placed where the `_a();` is?

I think [my comment below](https://github.com/microsoft/TypeScript/pull/29374#discussion_r478785677) might provide a better example. ES2015 constructors allow you to call `super()` anywhere in the constructor, such as like this:

```ts
class Derived extends Base {
  prop = 0;
  constructor(f) {
    if (f) {
      super(1);
    }
    else {
      console.log(super(2));
    }
  }
  toString() { return this.prop.toString(); }
}
```

If you were to inline the property assignment, you'd end up with duplication and even a runtime error in the `else` case above:

```js
// @target: es2015
class Derived extends Base {
  constructor(f) {
    if (f) {
      super(1);
      this.prop = 0; // duplication
    }
    else {
      console.log(super(2)); // error because `this.prop` is undefined when `toString()` is called.
      this.prop = 0; // duplication
    }
   toString() { return this.prop.toString(); }
  }
```

By wrapping the `super` call in an arrow, you can preserve the correct evaluation order (I'll use `_super_1` instead of `_a` for clarity below):

```js
// @target: es2015
class Derived extends Base {
  constructor(f) {
    let _super_1 = (...args) => {
      super(...args);
      this.prop = 0;
      return this;
    };
   
    if (f) {
      _super_1(1);
    }
    else {
      console.log(_super_1(2)); // no error because `this.prop` is defined.
    }
  }
  toString() { return this.prop.toString(); }
}
```

Obviously, downlevel to ES5 is a bit more verbose:
```js
// @target: es5
var Derived = /** @class */ (function (_super) {
    __extends(Derived, _super);
    function Derived(f) {
        var _this = this;
        var _super_1 = function () {
            _this = _super.apply(this, arguments) || this;
            _this.prop = true;
            return _this;
        };

        if (f) {
          _super_1(1);
        }
        else {
          console.log(_super_1(2));
        }
    }
    Derived.prototype.toString = function () { return this.prop.toString(); };
}(Base));
```

The question is whether we want to change the emit in this PR, or introduce a separate PR that allows `super()` calls anywhere in the constructor. I'm tempted to ask that we do this now, even for this case, so that the emit for existing code doesn't end up going through multiple changes as we release new versions.

@DanielRosenwasser do you have any thoughts here?